File: languages/runtime/ruby/backend_checker.rb
require 'json'
require 'stringio'

module BackendChecker
  class FailedTestcase
    attr_accessor :id, :input, :expected, :got, :stdOut

    def initialize(id, input, expected, got, std_out)
      @id = id
      @input = input
      @expected = expected
      @got = got
      @stdOut = std_out
    end

    def to_h
      {
        id: @id,
        input: @input,
        expected: @expected,
        got: @got,
        stdOut: @stdOut
      }
    end
  end

  class RunResult
    attr_accessor :isPassed, :failedTestcases, :memoryUsage, :executionTime, :stdOut, :stdErr

    def initialize
      @isPassed = true
      @failedTestcases = []
      @memoryUsage = 0.0
      @executionTime = 0.0
      @stdOut = ''
      @stdErr = ''
    end

    def to_h
      {
        isPassed: @isPassed,
        failedTestcases: @failedTestcases.map(&:to_h),
        memoryUsage: @memoryUsage,
        executionTime: @executionTime,
        stdOut: @stdOut,
        stdErr: @stdErr
      }
    end
  end

  def self.capture_stdout
    old_stdout = $stdout
    $stdout = StringIO.new
    
    result = yield
    
    output = $stdout.string
    $stdout = old_stdout
    
    { result: result, stdout: output }
  end

  def self.write_results(results, filename)
    File.write(filename, JSON.pretty_generate(results.to_h))
  rescue => e
    $stderr.puts "Failed to write results file."
    exit(1)
  end

  def self.run_tests(testcases, run_checker)
    start_time = Time.now
    
    results = run_checker.call(testcases)
    
    end_time = Time.now
    results.executionTime = ((end_time - start_time) * 1000).to_f
    
    write_results(results, 'results.json')
  end
end

File: languages/runtime/typescript/backend_checker.ts
import * as fs from 'fs';

export class FailedTestcase {
    id: number;
    input: any;
    expected: any;
    got: any;
    stdOut: string;

    constructor(id: number, input: any, expected: any, got: any, stdOut: string) {
        this.id = id;
        this.input = input;
        this.expected = expected;
        this.got = got;
        this.stdOut = stdOut;
    }
}

export class RunResult {
    isPassed: boolean = true;
    failedTestcases: FailedTestcase[] = [];
    memoryUsage: number = 0.0;
    executionTime: number = 0.0;
    stdOut: string = '';
    stdErr: string = '';
}

export function captureStdout<T>(fn: () => T): { result: T; stdout: string } {
    const originalLog = console.log;
    let output = '';
    
    console.log = (...args: any[]) => {
        output += args.join(' ') + '\n';
    };
    
    const result = fn();
    
    console.log = originalLog;
    
    return { result, stdout: output };
}

export function writeResults(results: RunResult, filename: string): void {
    try {
        fs.writeFileSync(filename, JSON.stringify(results, null, 2));
    } catch (error) {
        console.error('Failed to write results file.');
        process.exit(1);
    }
}

export function runTests<T>(testcases: T[], runChecker: (testcases: T[]) => RunResult): void {
    const startTime = Date.now();
    
    const results = runChecker(testcases);
    
    const endTime = Date.now();
    results.executionTime = endTime - startTime;
    
    writeResults(results, 'results.json');
}
